Amazon.QLDB.Driver.IntegrationTests/AsyncStatementExecutionTests.cs

```diff
@@ -0,0 +1,1053 @@
+                {
+                    var result = await txn.Execute(searchQuery, ionValue);
+
+                    IIonValue ionVal = null;
+                    await foreach (var row in result)
+                    {
+                        ionVal = row;
+                    }
+                    return ionVal;
+                });
+            }
+
+            if (searchResult.Type() != ionValue.Type())
+            {
+                Assert.Fail($"The queried value type, { searchResult.Type().ToString() }," +
+                    $"does not match { ionValue.Type().ToString() }.");
+            }
+        }
+
+        [TestMethod]
+        public async Task Execute_ExecuteLambdaThatDoesNotReturnValue_RecordIsUpdated()
+        {
+            // Given.
+            // Create Ion struct to insert.
+            IIonValue ionStruct = ValueFactory.NewEmptyStruct();
+            ionStruct.SetField(Constants.ColumnName, ValueFactory.NewString(Constants.SingleDocumentValue));
+
+            // When.
+            var query = $"INSERT INTO {Constants.TableName} ?";
+            await qldbDriver.Execute(async txn => await txn.Execute(query, ionStruct));
+
+            // Then.
+            var searchQuery = $@"SELECT VALUE {Constants.ColumnName} FROM {Constants.TableName}
+                                 WHERE {Constants.ColumnName} = '{Constants.SingleDocumentValue}'";
+            var value = await qldbDriver.Execute(async txn =>
+            {
+                var result = await txn.Execute(searchQuery);
+
+                string value = "";
+                await foreach (var row in result)
+                {
+                    value = row.StringValue;
+                }
+                return value;
+            });
+            Assert.AreEqual(Constants.SingleDocumentValue, value);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(BadRequestException))]
+        public async Task Execute_DeleteTableThatDoesntExist_ThrowsBadRequestException()
+        {
+            // Given.
+            var query = "DELETE FROM NonExistentTable";
+
+            // When.
+            await qldbDriver.Execute(async txn => await txn.Execute(query));
+        }
+
+        [TestMethod]
+        public async Task Execute_ExecutionMetrics()
+        {
+            await qldbDriver.Execute(async txn =>
+            {
+                var insertQuery = String.Format("INSERT INTO {0} << {{'col': 1}}, {{'col': 2}}, {{'col': 3}} >>",
+                    Constants.TableName);
+                await txn.Execute(insertQuery);
+            });
+
+            // Given
+            var selectQuery = String.Format("SELECT * FROM {0} as a, {0} as b, {0} as c, {0} as d, {0} as e, {0} as f",
+                Constants.TableName);
+
+            // When
+            await qldbDriver.Execute(async txn =>
+            {
+                var result = await txn.Execute(selectQuery);
+
+                await foreach (IIonValue row in result)
```

Comment:
Can we assert that result has records?

Currently if result is empty, this for loop won't do anything and we won't be able to assert the query stats.

This will need to be done on non-async side aswell

Reply:
If we want that, either we execute the query twice, or put a counter or boolean inside the for loop.  Because if I try to touch result, then I would void its 'iterativeness':
                var enumResult = result.GetAsyncEnumerator();
                assert(await enumResult.MoveNextAsync())  // now result cannot be used in the for loop anymore effectively.

Reply:
You can set 
```
var ioUsage
var timingInfo
```
outside of the iteration.

Then assert the values after the iteration

Reply:
Updated